aeppl/truncation.py

```diff
@@ -129,3 +144,216 @@ def censor_logprob(op, values, base_rv, lower_bound, upper_bound, **kwargs):
+    attempted when performing rejection sampling. An Error is raised if convergence is
+    not reached after that many steps.
+
+    TODO: Add Note about updates
+    """
+
+    if lower is None and upper is None:
+        raise ValueError("lower and upper cannot both be None")
+
+    if not (isinstance(rv.owner.op, RandomVariable) and rv.owner.op.ndim_supp == 0):
+        raise NotImplementedError(
+            f"truncation only implemented for univariate random variables, got {rv.owner.op}"
+        )
+
+    lower = at.as_tensor_variable(lower) if lower is not None else at.constant(-np.inf)
+    upper = at.as_tensor_variable(upper) if upper is not None else at.constant(np.inf)
+
+    if rng is None:
+        rng = shared(np.random.RandomState(), borrow=True)
+
+    # Try to use specialized Op
+    try:
+        return _truncated(rv.owner.op, lower, upper, rng, *rv.owner.inputs[1:])
+    except NotImplementedError:
+        pass
+
+    # Variables with `_` suffix identify dummy inputs for the OpFromGraph
+    graph_inputs = [rng, *rv.owner.inputs[1:], lower, upper]
+    graph_inputs_ = [inp.type() for inp in graph_inputs]
+    *rv_inputs_, lower_, upper_ = graph_inputs_
+
+    # Try to use inverted cdf sampling
+    try:
+        rv_ = rv.owner.op.make_node(*rv_inputs_).default_output()
+        rv_.name = rv.name
+        # For left truncated discrete RVs, we need to include the whole lower bound.
+        # This may result in draws below the truncation range, if any uniform == 0
+        lower_value = lower_ - 1 if rv.owner.op.dtype.startswith("int") else lower_
+        cdf_lower_ = at.exp(logcdf(rv_, lower_value))
+        cdf_upper_ = at.exp(logcdf(rv_, upper_))
+        uniform_ = at.random.uniform(
+            cdf_lower_,
+            cdf_upper_,
+            rng=rv_inputs_[0],
+            size=rv_inputs_[1],
+        )
+        truncated_rv_ = icdf(rv_, uniform_)
+        truncated_rv = TruncatedRV(
+            base_rv_op=rv.owner.op,
+            inputs=graph_inputs_,
+            outputs=[uniform_.owner.outputs[0], truncated_rv_],
+            inline=True,
+        )(*graph_inputs)
+        truncated_rv.update = (
+            truncated_rv.owner.inputs[0],
+            truncated_rv.owner.outputs[0],
+        )
+        return truncated_rv
+    except NotImplementedError:
+        pass
+
+    # Fallback to rejection sampling
+    # TODO: Handle potential broadcast by lower / upper
+
+    # Scan forces us to use a shared variable for the RNG
+    graph_inputs = graph_inputs[1:]
+    graph_inputs_ = graph_inputs_[1:]
+    *rv_inputs_, lower_, upper_ = (rng, *graph_inputs_)
+
+    rv_ = rv.owner.op.make_node(*rv_inputs_).default_output()
+
+    def loop_fn(truncated_rv, reject_draws, lower, upper, rng, *rv_inputs):
+        # We need to set default_update for scan to generate updates
+        next_rng, new_truncated_rv = rv.owner.op.make_node(rng, *rv_inputs).outputs
+        rng.default_update = next_rng
```

Comment:
TODO: Return updates manually instead of relying on default_update

Reply:
Yes, that's definitely needed before merging.